TLM/TLM/Manager/Impl/LaneConnectionManager.cs

```diff
@@ -84,44 +99,101 @@ public class LaneConnectionManager
             throw new Exception($"Unreachable code. sourceLaneId:{sourceLaneId}, segmentId:{segmentId} ");
         }
 
-        public bool HasConnections(uint sourceLaneId) {
+        public bool HasOutgoingConnections(uint sourceLaneId) {
             if (!Options.laneConnectorEnabled) {
                 return false;
             }
-            return HasConnections(sourceLaneId, IsHeadingTowardsStartNode(sourceLaneId));
+
+            if(!ValidateLane(sourceLaneId))
+                return false;
+
+            return HasOutgoingConnections(sourceLaneId, IsHeadingTowardsStartNode(sourceLaneId));
         }
 
         /// <summary>
         /// Determines if the given lane has outgoing connections
         /// </summary>
         /// <param name="sourceLaneId"></param>
         /// <returns></returns>
-        public bool HasConnections(uint sourceLaneId, bool startNode) {
+        public bool HasOutgoingConnections(uint sourceLaneId, bool startNode) {
             if (!Options.laneConnectorEnabled) {
                 return false;
             }
 
-            int nodeArrayIndex = startNode ? 0 : 1;
+            if(!ValidateLane(sourceLaneId))
+                return false;
 
-            return Flags.laneConnections[sourceLaneId] != null &&
-                   Flags.laneConnections[sourceLaneId][nodeArrayIndex] != null;
+            var targets = connections_.GetConnections(sourceLaneId, startNode);
+            return targets != null && targets.Length > 0;
+        }
+
+        /// <summary>
+        /// Determines if the given lane has incoming/outgoing connections
+        /// </summary>
+        public bool HasConnections(uint laneId, ushort nodeId) {
+            bool startNode = laneId.ToLane().IsStartNode(nodeId);
+            if(HasOutgoingConnections(laneId, startNode))
+                return true; // outgoing connection
+
+            //return false;
+            return HasIncommingConnections(laneId, nodeId);
+        }
+
+        public bool HasIncommingConnections(uint targetLaneId, ushort nodeId) {
+            if(!ValidateLane(targetLaneId))
+                return false;
+
+            ref NetNode node = ref nodeId.ToNode();
+            for(int segmentIndex = 0; segmentIndex < 8; ++segmentIndex) {
+                ushort segmentId = node.GetSegment(segmentIndex);
+                if(segmentId == 0)
+                    continue;
+
+                ref NetSegment segment = ref segmentId.ToSegment();
+                if(!segment.IsValid()) {
+                    HandleInvalidSegmentImpl(segmentId);
```

Comment:
Should it also then `return false`?

Reply:
even if segment is invalid I still need to remove invalid lanes.

Reply:
Yup, the invalid lanes need removing - I thought that's what `HandleInvalidSegmentImpl()` did?

After that shoudln't it `continue` out of this iteration of the `for (segmentIndex ... )` loop?

What in the remainder of the method block (L161-L177) would remove invalid lanes?

Reply:
I check for backward connections to the segment too. now that the connections are one-way things are a bit more complicated.